src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGeneratorImpl.scala

```diff
@@ -0,0 +1,184 @@
+
+  protected def prepareHeader(
+    blockNumber: BigInt, parent: Block,
+    beneficiary: Address, blockTimestamp: Long,
+    ommers: X
+  ): BlockHeader
+
+  protected def prepareBlock(
+    parent: Block,
+    transactions: Seq[SignedTransaction],
+    beneficiary: Address,
+    blockNumber: BigInt,
+    blockPreparator: BlockPreparator,
+    ommers: X
+  ): PendingBlockAndState = {
+
+    val blockTimestamp = blockTimestampProvider.getEpochSecond
+    val header = prepareHeader(blockNumber, parent, beneficiary, blockTimestamp, ommers)
+    val transactionsForBlock = prepareTransactions(transactions, header.gasLimit)
+    val body = newBlockBody(transactionsForBlock, ommers)
+    val block = Block(header, body)
+
+    val prepared = blockPreparator.prepareBlock(block) match {
+      case BlockPreparationResult(prepareBlock, BlockResult(_, gasUsed, receipts), stateRoot, updatedWorld) =>
+        val receiptsLogs: Seq[Array[Byte]] = BloomFilter.EmptyBloomFilter.toArray +: receipts.map(_.logsBloomFilter.toArray)
+        val bloomFilter = ByteString(or(receiptsLogs: _*))
+
+        PendingBlockAndState(PendingBlock(block.copy(header = block.header.copy(
+          transactionsRoot = buildMpt(prepareBlock.body.transactionList, SignedTransaction.byteArraySerializable),
+          stateRoot = stateRoot,
+          receiptsRoot = buildMpt(receipts, Receipt.byteArraySerializable),
+          logsBloom = bloomFilter,
+          gasUsed = gasUsed),
+          body = prepareBlock.body), receipts), updatedWorld)
+    }
+    prepared
+  }
+
+  protected def prepareTransactions(
+    transactions: Seq[SignedTransaction],
+    blockGasLimit: BigInt
+  ): immutable.Seq[SignedTransaction] = {
+
+    val sortedTransactions: immutable.Seq[SignedTransaction] = transactions
+      .groupBy(_.senderAddress).values.toList
+      .flatMap { txsFromSender =>
+        val ordered = txsFromSender
+          .sortBy(-_.tx.gasPrice)
+          .sortBy(_.tx.nonce)
+          .foldLeft(Seq.empty[SignedTransaction]) { case (txs, tx) =>
+            if (txs.exists(_.tx.nonce == tx.tx.nonce)) {
+              txs
+            } else {
+              txs :+ tx
+            }
+          }
+          .takeWhile(_.tx.gasLimit <= blockGasLimit)
+        ordered.headOption.map(_.tx.gasPrice -> ordered)
+      }
+      .sortBy { case (gasPrice, _) => gasPrice }
+      .reverse
+      .flatMap { case (_, txs) => txs }
+
+    val transactionsForBlock: immutable.Seq[SignedTransaction] = sortedTransactions
+      .scanLeft(BigInt(0), None: Option[SignedTransaction]) { case ((accumulatedGas, _), stx) => (accumulatedGas + stx.tx.gasLimit, Some(stx)) }
+      .collect { case (gas, Some(stx)) => (gas, stx) }
+      .takeWhile { case (gas, _) => gas <= blockGasLimit }
+      .map { case (_, stx) => stx }
+
+    transactionsForBlock
+  }
+
+  //returns maximal limit to be able to include as many transactions as possible
+  protected def calculateGasLimit(parentGas: BigInt): BigInt = {
```

Comment:
I think it's ethash-specific, isn't it?

Reply:
True, it's used there. It just seems like a leaky abstraction - we have _generic_ skeleton for BlockGenerator, which contains ethash (subclass) specific stuff.

Reply:
I'm not sure if `gasLimit` is ethash- or Pow-specific. At any rate you might argue than _in general_ consensuses have a gas limit, and specific one may choose to ignore it (in practice setting it to a huge number)